### PR TITLE
Bratseth/remove jobs for removed deployments

### PR DESCRIPTION
**Pull request by bratseth**
@jonmv please review

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 1]
```diff
@@ -345,74 +346,60 @@ private boolean deploysTo(Application application, JobType jobType) {
...
                                       application, cause));
             return application;
         }
 
-        // Ignore applications that are not associated with a project
-        if ( ! application.deploymentJobs().projectId().isPresent()) {
-            return application;
-        }
-
+        if ( ! force && ! allowedTriggering(jobType, application)) return application;
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: Move the 
```java
if (jobType == null) return ...
```
outside, lest you'll have NPEs below.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentMetricsMaintainer.java [hunk 2]
```diff
@@ -51,8 +53,10 @@ protected void maintain() {
                     }
                 }
                 catch (UncheckedIOException e) {
-                    log.log(Level.WARNING, "Failed talking to YAMAS: " + e.getMessage() +
-                            ", retrying in " + maintenanceInterval());
+                    if ( ! hasWarned) // produce only one warning per maintenance interval
+                        log.log(Level.WARNING, "Failed talking to YAMAS: " + Exceptions.toMessageString(e) +
+                                               ". Retrying in " + maintenanceInterval());
+                    hasWarned = true;
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentMetricsMaintainer.java]: Forgot to add this to the issue tracking this logging, but this warning was logged roughly once per 20 minutes on the controller yesterday, so it was already ~ once per run . This formatting is better, though.

**Review by jonmv — CHANGES_REQUESTED** [review]
Fix the NPE, then go :)

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java (hunk 1 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: That can not actually happen, but I agree it is clearer that way. Fixing ...

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentMetricsMaintainer.java (hunk 2 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentMetricsMaintainer.java]: No, it is clear from the code it is not. And from this morning the log contains

{code}
[2017-10-26 09:02:38.190] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.190] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.190] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.191] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.191] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.191] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.191] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.192] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.192] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.192] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.192] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.193] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.193] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.193] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.194] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.194] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.194] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.194] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.195] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.195] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.195] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.196] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.196] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.196] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.196] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.197] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.197] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.197] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.197] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.198] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed talking to YAMAS: java.net.UnknownHostException: yamas.ops.yahoo.com, retrying in PT10M
[2017-10-26 09:02:38.198] WARNING : container        Container.com.yahoo.vespa.hosted.controller.maintenance.DeploymentMetricsMaintainer	Failed 
....
{code}

Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentMetricsMaintainer.java]: I mean it was, in practice, once per run. Yesterday. Today: ouch! Change definitely warranted, then.

**Pull request merged**